File: app/src/main/java/org/nv95/openmanga/utils/network/NetworkUtils.java
package org.nv95.openmanga.utils.network;

import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.ProxyInfo;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import org.apache.http.conn.ssl.SSLSocketFactory;
import org.json.JSONException;
import org.json.JSONObject;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.nv95.openmanga.legacy.items.RESTResponse;
import org.nv95.openmanga.legacy.utils.CookieParser;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;
import java.net.HttpURLConnection;
import java.net.URLEncoder;
import java.util.concurrent.TimeUnit;

import javax.net.ssl.HttpsURLConnection;

import info.guardianproject.netcipher.NetCipher;
import info.guardianproject.netcipher.client.*;
import info.guardianproject.netcipher.proxy.OrbotHelper;
import info.guardianproject.netcipher.proxy.TorServiceUtils;
import okhttp3.CacheControl;
import okhttp3.Headers;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;

/**
 * Created by nv95 on 29.11.16.
 */

public class NetworkUtils {

	private static final CacheControl CACHE_CONTROL_DEFAULT = new CacheControl.Builder().maxAge(10, TimeUnit.MINUTES).build();
	private static final Headers HEADERS_DEFAULT = new Headers.Builder().build();

	private static OkHttpClient sHttpClient = null;

	@NonNull
	private static OkHttpClient.Builder getClientBuilder() {
		return new OkHttpClient.Builder()
				.addInterceptor(new CloudflareInterceptor());
	}

	public static void init(Context context, boolean useTor) {
		OkHttpClient.Builder builder = getClientBuilder();
		if (useTor && OrbotHelper.get(context).init()) {
			try {
				StrongOkHttpClientBuilder.forMaxSecurity(context)
						.applyTo(builder, new Intent()
								.putExtra(OrbotHelper.EXTRA_STATUS, "ON")); //TODO wtf
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
		sHttpClient = builder.build();
	}

	@NonNull
	public static String getString(@NonNull String url) throws IOException {
		return getString(url, HEADERS_DEFAULT);
	}

	@NonNull
	public static String getString(@NonNull String url, @NonNull Headers headers) throws IOException {
		Request.Builder builder = new Request.Builder()
				.url(url)
				.headers(headers)
				.cacheControl(CACHE_CONTROL_DEFAULT)
				.get();
		Response response = null;
		try {
			response = sHttpClient.newCall(builder.build()).execute();
			ResponseBody body = response.body();
			if (body == null) {
				throw new IOException("ResponseBody is null");
			} else {
				return body.string();
			}
		} finally {
			if (response != null) {
				response.close();
			}
		}
	}

	@NonNull
	public static Document getDocument(@NonNull String url) throws IOException {
		return getDocument(url, HEADERS_DEFAULT);
	}

	@NonNull
	public static Document getDocument(@NonNull String url, @NonNull Headers headers) throws IOException {
		return Jsoup.parse(getString(url, headers), url);
	}

	@NonNull
	public static JSONObject getJSONObject(@NonNull String url) throws IOException, JSONException {
		return new JSONObject(getString(url));
	}

	@Deprecated
	public static Document httpPost(@NonNull String url, @Nullable String cookie, @Nullable String[] data) throws IOException {
		InputStream is = null;
		try {
			HttpURLConnection con = createConnection(url);
			con.setRequestMethod("POST");
			if (!TextUtils.isEmpty(cookie)) {
				con.setRequestProperty("Cookie", cookie);
			}
			if (data != null) {
				con.setDoOutput(true);
				DataOutputStream out = new DataOutputStream(con.getOutputStream());

				out.writeBytes(makeQuery(data));
				out.flush();
				out.close();
			}
			is = con.getInputStream();
			return Jsoup.parse(is, con.getContentEncoding(), url);
		} finally {
			if (is != null) {
				is.close();
			}
		}
	}

	@Deprecated
	@NonNull
	public static String getRaw(@NonNull String url, @Nullable String cookie) throws IOException {
		BufferedReader reader = null;
		try {
			HttpURLConnection con = createConnection(url);
			if (!TextUtils.isEmpty(cookie)) {
				con.setRequestProperty("Cookie", cookie);
			}
			reader = new BufferedReader(new InputStreamReader(con.getInputStream()));
			StringBuilder out = new StringBuilder();
			String line;
			while ((line = reader.readLine()) != null) {
				out.append(line);
			}
			return out.toString();
		} finally {
			if (reader != null) {
				reader.close();
			}
		}
	}

	@Deprecated
	@Nullable
	public static CookieParser authorize(String url, String... data) {
		DataOutputStream out = null;
		try {
			HttpURLConnection con = createConnection(url);
			con.setRequestMethod("POST");
			con.setDoOutput(true);
			con.setInstanceFollowRedirects(true);
			out = new DataOutputStream(con.getOutputStream());
			out.writeBytes(makeQuery(data));
			out.flush();
			con.connect();
			if (con.getResponseCode() == HttpURLConnection.HTTP_OK) {
				return new CookieParser(con.getHeaderFields().get("Set-Cookie"));
			} else {
				return null;
			}
		} catch (Exception e) {
			e.printStackTrace();
			return null;
		} finally {
			try {
				if (out != null) {
					out.close();
				}
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}

	@Deprecated
	public static RESTResponse restQuery(String url, @Nullable String token, String method, String... data) {
		BufferedReader reader = null;
		try {
			HttpURLConnection con = createConnection("GET".equals(method) ? url + "?" + makeQuery(data) : url);
			if (!TextUtils.isEmpty(token)) {
				con.setRequestProperty("X-AuthToken", token);
			}
			con.setRequestMethod(method);
			if (!"GET".equals(method)) {
				con.setDoOutput(true);
				DataOutputStream out = new DataOutputStream(con.getOutputStream());
				out.writeBytes(NetworkUtils.makeQuery(data));
				out.flush();
				out.close();
			}
			int respCode = con.getResponseCode();
			reader = new BufferedReader(new InputStreamReader(isOk(respCode) ? con.getInputStream() : con.getErrorStream()));
			StringBuilder out = new StringBuilder();
			String line;
			while ((line = reader.readLine()) != null) {
				out.append(line);
			}
			return new RESTResponse(new JSONObject(out.toString()), respCode);
		} catch (Exception e) {
			e.printStackTrace();
			return RESTResponse.fromThrowable(e);
		} finally {
			if (reader != null) {
				try {
					reader.close();
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
		}
	}

	@Deprecated
	@NonNull
	private static String makeQuery(@NonNull String[] data) throws UnsupportedEncodingException {
		StringBuilder query = new StringBuilder();
		for (int i = 0; i < data.length; i = i + 2) {
			query.append(URLEncoder.encode(data[i], "UTF-8")).append("=").append(URLEncoder.encode(data[i + 1], "UTF-8")).append("&");
		}
		if (query.length() > 1) {
			query.deleteCharAt(query.length() - 1);
		}
		return query.toString();
	}

	@Deprecated
	private static boolean isOk(int responseCode) {
		return responseCode >= 200 && responseCode < 300;
	}

	@Deprecated
	private static HttpURLConnection createConnection(String url) throws IOException {
		HttpURLConnection con = NetCipher.getHttpURLConnection(url);
		if (con instanceof HttpsURLConnection) {
			((HttpsURLConnection) con).setSSLSocketFactory(NoSSLv3SocketFactory.getInstance());
			((HttpsURLConnection) con).setHostnameVerifier(SSLSocketFactory.ALLOW_ALL_HOSTNAME_VERIFIER);
		}
		con.setConnectTimeout(15000);
		return con;
	}
}

File: app/src/main/java/org/nv95/openmanga/core/network/NetworkUtils.kt
package org.nv95.openmanga.core.network

import android.content.Context
import com.itkacher.okhttpprofiler.OkHttpProfilerInterceptor
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONException
import org.json.JSONObject
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.nv95.openmanga.BuildConfig
import org.nv95.openmanga.core.sources.ConnectionSource
import org.nv95.openmanga.di.KoinJavaComponent
import org.nv95.openmanga.items.RESTResponse
import timber.log.Timber
import java.io.IOException
import java.io.UnsupportedEncodingException
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

/**
 * Created by nv95 on 29.11.16.
 */

object NetworkUtils {

	private const val TAG = "NetworkUtils"
	const val HTTP_GET = "GET"
	const val HTTP_POST = "POST"
	const val HTTP_PUT = "PUT"
	const val HTTP_DELETE = "DELETE"

	private val client by lazy {
		OkHttpClient.Builder()
				.apply {
					if (BuildConfig.DEBUG) addInterceptor(OkHttpProfilerInterceptor())
				}
				.followSslRedirects(true)
				.followRedirects(true)
				.connectTimeout(15, TimeUnit.SECONDS)
				.addInterceptor(UserAgentInterceptor())
				.build()
	}

	@JvmStatic
	@Throws(IOException::class)
	fun httpGet(url: String, cookie: String? = null): Document {
		val request = Request.Builder()
				.url(url)
				.get()
		if (!cookie.isNullOrBlank()) {
			request.header("Cookie", cookie)
		}
		return client.newCall(request.build()).execute().use { response ->
			Jsoup.parse(response.body()?.string(), url)
		}
	}

	@Deprecated("")
	@JvmStatic
	@Throws(IOException::class)
	fun httpPost(url: String, cookie: String?, data: Array<String>) =
			httpPost(url, cookie, data.toList().zipWithNext { a, b -> a to b }.toMap())

	@JvmStatic
	@Throws(IOException::class)
	fun httpPost(url: String, cookie: String? = null, data: Map<String, String>? = null): Document {
		val body = FormBody.Builder()
		data?.entries?.forEach { x -> body.addEncoded(x.key, x.value) }
		val request = Request.Builder()
				.url(url)
				.post(body.build())
		if (!cookie.isNullOrBlank()) {
			request.header("Cookie", cookie)
		}
		return client.newCall(request.build()).execute().use { response ->
			Jsoup.parse(response.body()?.string(), url)
		}
	}

	@JvmStatic
	@Throws(IOException::class)
	fun getRaw(url: String, cookie: String? = null): String {
		val request = Request.Builder()
				.url(url)
				.get()
		if (!cookie.isNullOrBlank()) {
			request.header("Cookie", cookie)
		}
		return client.newCall(request.build()).execute().use { response ->
			response.body()?.string() ?: ""
		}
	}

	@JvmStatic
	@Throws(IOException::class, JSONException::class)
	fun getJsonObject(url: String): JSONObject {
		return JSONObject(getRaw(url, null))
	}

	@JvmStatic
	@Deprecated("")
	fun authorize(url: String, vararg data: String) = authorize(url, data.toList().zipWithNext { a, b -> a to b }.toMap())

	@JvmStatic
	fun authorize(url: String, data: Map<String, String>): CookieParser? = try {
		val body = FormBody.Builder()
		data.entries.forEach { x -> body.addEncoded(x.key, x.value) }
		val request = Request.Builder()
				.url(url)
				.post(body.build())
		client.newCall(request.build()).execute().use { response ->
			CookieParser(response.headers("Set-Cookie"))
		}
	} catch (e: Exception) {
		Timber.tag(TAG).e(e)
		null
	}

	@JvmStatic
	fun restQuery(url: String, token: String?, method: String, vararg data: String) =
			restQuery(url, token, method, data.toList().zipWithNext { a, b -> a to b }.toMap())

	@JvmStatic
	fun restQuery(url: String, token: String?, method: String, data: Map<String, String>): RESTResponse {
		try {
			val request = Request.Builder()
			when (method) {
				HTTP_GET -> request.url(url + "?" + makeQuery(data)).get()
				else -> {
					val body = FormBody.Builder()
					data.entries.forEach { x -> body.addEncoded(x.key, x.value) }
					request.url(url).method(method, body.build())
				}
			}
			if (!token.isNullOrBlank()) {
				request.header("X-AuthToken", token)
			}
			return client.newCall(request.build()).execute().use { response ->
				return RESTResponse(JSONObject(response.body()!!.string()), response.code())
			}
		} catch (e: Exception) {
			Timber.tag(TAG).e(e)
			return RESTResponse.fromThrowable(e)
		}
	}

	private fun makeQuery(data: Map<String, String>) = data.entries.joinToString("&") { x ->
		URLEncoder.encode(x.key, "UTF-8") + "=" + URLEncoder.encode(x.value, "UTF-8")
	}

	@Deprecated("")
	@Throws(UnsupportedEncodingException::class)
	private fun makeQuery(data: Array<out String>): String {
		val query = StringBuilder()
		var i = 0
		while (i < data.size) {
			query.append(URLEncoder.encode(data[i], "UTF-8")).append("=").append(URLEncoder.encode(data[i + 1], "UTF-8")).append("&")
			i += 2
		}
		if (query.length > 1) {
			query.deleteCharAt(query.length - 1)
		}
		val queryString = query.toString()
		return queryString
	}

	/**
	 * Use [ConnectionSource.isConnectionAvailable]
	 */
	@Deprecated("")
	@JvmStatic
	fun checkConnection(context: Context): Boolean {
		return KoinJavaComponent.get(ConnectionSource::class.java).isConnectionAvailable()
	}

	/**
	 * Use [ConnectionSource.isConnectionAvailable]
	 */
	@Deprecated("")
	@JvmStatic
	fun checkConnection(context: Context, onlyWiFi: Boolean): Boolean {
		return KoinJavaComponent.get(ConnectionSource::class.java).isConnectionAvailable(onlyWiFi)
	}

	@JvmStatic
	fun getHttpClient() = client
}
